File: Structural/Proxy/C#/ProxyExample/ProxyExample/SubjectProxy.cs
namespace ProxyExample;

internal class SubjectProxy : IRealSubject
{
	private IRealSubject? _realSubject;

	public string GetData()
	{
		if (_realSubject == null)
			_realSubject = new RealSubject();

		return _realSubject.GetData();
	}
}

File: Structural/Proxy/Delphi/SubjectProxy.pas
unit SubjectProxy;

interface

uses
  RealSubject, RealSubjectImp;

type
  TSubjectProxy = class(TInterfacedObject, IRealSubject)
  private
    FRealSubject: IRealSubject;
  public
    constructor Create();
    destructor Destroy(); override;
    function GetData(): string;
  end;

implementation

constructor TSubjectProxy.Create();
begin
  inherited Create();
  FRealSubject := nil;
end;

destructor TSubjectProxy.Destroy();
begin
  if Assigned(FRealSubject) then
    FRealSubject._Release;
  inherited Destroy();
end;

function TSubjectProxy.GetData(): string;
begin
  if not Assigned(FRealSubject) then
    FRealSubject := TRealSubject.Create as IRealSubject;
  Result := FRealSubject.GetData();
end;

end.
